ksql-engine/src/main/java/io/confluent/ksql/function/udaf/array/CollectListUdaf.java

```diff
@@ -0,0 +1,174 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.array;
+
+import avro.shaded.com.google.common.collect.Lists;
+import io.confluent.ksql.function.udaf.TableUdaf;
+import io.confluent.ksql.function.udaf.Udaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import java.util.List;
+
+@UdafDescription(name = "collect_list", 
+    description = "Gather all of the grouped values into a single Array field")
+public final class CollectListUdaf {
+
+  private CollectListUdaf() {
+    // just to make the checkstyle happy
+  }
+  
+  @UdafFactory(description = "collect values of a Bigint field into a single Array")
+  public static Udaf<Long, List<Long>> createCollectListLong() {
+    return new TableUdaf<Long, List<Long>>() {
+      @Override
+      public List<Long> undo(final Long valueToUndo, final List<Long> aggregateValue) {
+        aggregateValue.remove(aggregateValue.lastIndexOf(valueToUndo));
+        return aggregateValue;
+      }
+
+      @Override
+      public List<Long> initialize() {
+        return Lists.newArrayList();
+      }
+
+      @Override
+      public List<Long> aggregate(final Long thisValue, final List<Long> aggregate) {
+        aggregate.add(thisValue);
+        return aggregate;
+      }
+
+      @Override
+      public List<Long> merge(final List<Long> aggOne, final List<Long> aggTwo) {
+        aggOne.addAll(aggTwo);
+        return aggOne;
+      }
+    };
+  }
+
+  @UdafFactory(description = "collect values of an Integer field into a single Array")
+  public static Udaf<Integer, List<Integer>> createCollectListInt() {
+    return new TableUdaf<Integer, List<Integer>>() {
+      @Override
+      public List<Integer> undo(final Integer valueToUndo, final List<Integer> aggregateValue) {
+        aggregateValue.remove(aggregateValue.lastIndexOf(valueToUndo));
+        return aggregateValue;
+      }
+
+      @Override
+      public List<Integer> initialize() {
+        return Lists.newArrayList();
+      }
+
+      @Override
+      public List<Integer> aggregate(final Integer thisValue, final List<Integer> aggregate) {
+        aggregate.add(thisValue);
+        return aggregate;
+      }
+
+      @Override
+      public List<Integer> merge(final List<Integer> aggOne, final List<Integer> aggTwo) {
+        aggOne.addAll(aggTwo);
+        return aggOne;
+      }
+    };
+  }
+
+  @UdafFactory(description = "collect values of a Double field into a single Array")
+  public static Udaf<Double, List<Double>> createCollectListDouble() {
```

Comment:
Just a thought... not sure if its supported or not, but can you achieve the same with a single UADF of type `Number`, rather than different derived number types?  Maybe...

Even if you can't, you should be able to write a single generic implementation and then typed versions from each method, e.g.

```
@UdafFactory(description = "collect values of a BigInt field into a single Array")
public static Udaf<Long, List<Long>> createCollectListLong() {
   return new CollectList<>();
}


@UdafFactory(description = "collect values of a Double field into a single Array")
public static Udaf<Double, List<Double>> createCollectListDouble() {
   return new CollectList<>();
}

private static final class ColllectList<T extends Number> implements TableUdaf<T, List<T>> {
      @Override
      public List<T> undo(final T valueToUndo, final List<T> aggregateValue) {
        aggregateValue.remove(aggregateValue.lastIndexOf(valueToUndo));
        return aggregateValue;
      }
       @Override
      public List<T> initialize() {
        return Lists.newArrayList();
      }
       @Override
      public List<T> aggregate(final T thisValue, final List<T> aggregate) {
        aggregate.add(thisValue);
        return aggregate;
      }
       @Override
      public List<T> merge(final List<T> aggOne, final List<T> aggTwo) {
        aggOne.addAll(aggTwo);
        return aggOne;
      }
    };
```

Reply:
frankly this was a source of much frustration while coming up with these 2 UDAFs (although that's kind of the point of doing a couple as a test run off what we expect the rest of the world to do, kinda like we gathered up the feedback on building UDFs).   From the developer perspective, it feels like there's way too much boilerplate here for very little "real" code, given the scaffolding that's enforced by the UDAF-invoking framework. I quickly found myself wanteing to just write a single one of each of merge/aggregate/undo that works on Object!  (this won't work because of the way the UDF/UDAF invooker tries to bind to very precise type signatures for the methods, which is also a big part of the reason we can't write functions on Map/List/Struct types, which is going to be very frustrating too).

Reply:
that said, i'll give your suggestion a whirl, see what happens

Reply:
This also has the advantage that we don't leave the other types untested by the unit-test. That said, if you can't get it to work you can at least do something similar for the unit tests so that we run tests for all the types. The sum UDAF does something similar.

Reply:
Type safety is a good think @blueedgenick ;)

Reply:
thinking about a bit more revealed that a version of your genericised suggestion would work just fine so i've gone ahead and modified both these udafs to use it - how do you think ?